openrelik_worker_common/mount_utils.py

```diff
@@ -102,13 +133,129 @@ def _losetup(self) -> str:
+            self.dlm = self._get_redlockmanager()
+            lock = self.dlm.lock(
+                f"{hostname}-{devname}", self.LOCK_TIMEOUT_MILLISECONDS
+            )
+            if lock:
+                self.redlock = lock
+                logger.info(f"Redlock succesfully set: {lock.resource}")
+                return devname
+
+        raise RuntimeError("Error getting free NBD device: All NBD devices locked!")
+
+    def _nbdsetup(self):
+        """Map QCOW image file to NBD device using qemu-nbd and probe partitions.
+
+        Returns:
+            str: block device created by qemu-nbd
+
+        Raises:
+            RuntimeError: if there was an error running qemu-nbd or fdisk.
+        """
+        # Get and lock a free nbd device
+        self.blkdevice = self._get_free_nbd_device()
+        nbdsetup_command = [
+            "sudo",
+            "qemu-nbd",
+            "--read-only",
+            "--connect",
+            self.blkdevice,
+            self.image_path,
+        ]
+
+        process = subprocess.run(
+            nbdsetup_command, capture_output=True, check=False, text=True
+        )
+        if process.returncode == 0:
+            logger.info(
+                f"qemu-nbd: success creating {self.blkdevice} for {self.image_path}"
+            )
+        else:
+            logger.error(
+                f"qemu-nbd: failed creating {self.blkdevice} for {self.image_path}: {process.stderr} {process.stdout}"
+            )
+            raise RuntimeError(
+                f"Error running qemu-nbd: {process.stderr} {process.stdout}"
+            )
+
+        # This sleep is needed for qemu-nbd to activate the nbd device
+        time.sleep(0.5)
+
+        # Probe partitions with fdisk
+        fdisk_command = [
+            "sudo",
+            "fdisk",
+            "-l",
+            self.blkdevice.strip(),
+        ]
+
+        process = subprocess.run(
+            fdisk_command, capture_output=True, check=False, text=True
+        )
+        if process.returncode == 0:
+            logger.info(
+                f"fdisk: success probing {self.blkdevice} for {self.image_path}"
+            )
+        else:
+            logger.error(
+                f"fdisk: failed probing {self.blkdevice} for {self.image_path}: {process.stderr} {process.stdout}"
+            )
+            raise RuntimeError(
+                f"Error fdisk: failed probing: {process.stderr} {process.stdout}"
+            )
+
+        return self.blkdevice
+
     def _required_tools_available(self) -> bool:
         """Check if required cli tools are available.
 
+        Required tools can be installed on Debian by adding apt installing the
+        following packages:
+        * fdisk
+        * qemu-utils
+
         Returns:
             tuple: tuple of return bool and error message
         """
-        tools = ["lsblk", "blkid", "mount"]
+        tools = ["lsblk", "blkid", "mount", "qemu-nbd", "sudo"]
```

Comment:
Why do we need "sudo"? The container should run as root to make mounting work anyways. Or is there some other reason I'm missing? :)

Reply:
Dejavu ;) - https://github.com/openrelik/openrelik-worker-common/pull/21#discussion_r1965533265

Reply:
Haha, lol. My memory is good but short :)